File: medium/1657. Determine if Two Strings Are Close/1657. Determine if Two Strings Are Close.go
package main

import "sort"

func closeStrings(word1 string, word2 string) bool {
	if len(word1) != len(word2) {
		return false
	}

	d1 := make(map[rune]int)
	d2 := make(map[rune]int)

	for i := 0; i < len(word1); i++ {
		d1[rune(word1[i])] += 1
		d2[rune(word2[i])] += 1
	}

	a1 := []int{}
	a2 := []int{}

	for d := range d1 {
		if _, v := d2[d]; !v {
			return false
		}

		a1 = append(a1, d1[d])
	}

	for d := range d2 {
		if _, v := d1[d]; !v {
			return false
		}

		a2 = append(a2, d2[d])
	}

	sort.Ints(a1)
	sort.Ints(a2)

	for i := 0; i < len(a1); i++ {
		if a1[i] != a2[i] {
			return false
		}
	}

	return true
}

File: medium/1657. Determine if Two Strings Are Close/1657. Determine if Two Strings Are Close.py
from collections import Counter


class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        if len(word1) != len(word2):
            return False

        # if set(word1) != set(word2):
        #     return False

        # w1_dict = defaultdict(int)
        # for w in word1:
        #     w1_dict[w] += 1

        # w2_dict = defaultdict(int)
        # for w in word2:
        #     w2_dict[w] += 1

        # w1_count = [i for _, i in w1_dict.items()]
        # w2_count = [i for _, i in w2_dict.items()]

        # w1_count.sort()
        # w2_count.sort()

        # for i in range(len(w1_count)):
        #     if w1_count[i] != w2_count[i]:
        #         return False

        # return True

        d1 = Counter(word1)
        d2 = Counter(word2)

        if set(d1.keys()) != set(d2.keys()):
            return False

        return sorted(d1.values()) == sorted(d2.values())
